juice/src/net/common/rnn.rs

```diff
@@ -0,0 +1,156 @@
+use std::cell::RefCell;
+use std::fmt::{Debug, Formatter};
+use std::rc::Rc;
+
+use crate::co::{IBackend, ITensorDesc};
+use crate::net::{Context, Descriptor, Layer, LayerError, LearnableParams};
+use crate::util::LayerOps;
+use crate::weight::FillerType;
+use coaster::SharedTensor;
+use conn::{DirectionMode, RnnAlgorithm, RnnInputMode, RnnNetworkMode};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct RnnConfig {
+    /// Size of the hidden layer.
+    pub hidden_size: usize,
+    /// Number of hidden layers.
+    pub num_layers: usize,
+    /// Type of RNN.
+    pub rnn_type: RnnNetworkMode,
+    /// Dropout probability.
+    pub dropout_probability: f32,
+    /// Dropout seed
+    pub dropout_seed: u64,
+    /// Input mode.
+    pub input_mode: RnnInputMode,
+    /// RNN direction.
+    pub direction_mode: DirectionMode,
+}
+
+pub struct Rnn<B: conn::Rnn<f32>> {
+    descriptor: Descriptor,
+    config: RnnConfig,
+
+    // RNN config (batch size agnostic).
+    rnn_context: RefCell<B::CRNN>,
+
+    // RNN weights containing linear weights and biases for the RNN units.
+    weights: Rc<RefCell<LearnableParams>>,
+}
+
+impl<B: conn::Rnn<f32>> Rnn<B> {
+    pub fn new(backend: &B, mut descriptor: Descriptor, config: &RnnConfig) -> Self {
```

Comment:
nit: Could we consider making this failable, at the call site we already have failable initializations

Reply:
Done.